### PR TITLE
[SPARK-46539][SQL] SELECT * EXCEPT(all fields from a struct) results in an assertion failure

### PR DESCRIPTION
**Pull request by stefankandic**
### What changes were proposed in this pull request?

Fixing the assertion error which occurs when we do SELECT .. EXCEPT(every field from a struct) by adding a check for an empty struct

### Why are the changes needed?

Because this is a valid query that should just return an empty struct rather than fail during serialization.

### Does this PR introduce _any_ user-facing change?

Yes, users should no longer see this error and instead get an empty struct '{ }'

### How was this patch tested?

By adding new UT to existing selectExcept tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 1]
```diff
@@ -325,11 +325,16 @@ case class ExpressionEncoder[T](
   assert(serializer.forall(_.references.isEmpty), "serializer cannot reference any attributes.")
   assert(serializer.flatMap { ser =>
     val boundRefs = ser.collect { case b: BoundReference => b }
-    assert(boundRefs.nonEmpty,
+    assert(boundRefs.nonEmpty || isEmptyStruct(ser),
       "each serializer expression should contain at least one `BoundReference`")
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:329]: Please, update the assert message according to your changes.

**Comment by MaxGekk** [comment]
@milastdbx Are you ok with the changes?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 2]
```diff
@@ -325,11 +325,17 @@ case class ExpressionEncoder[T](
   assert(serializer.forall(_.references.isEmpty), "serializer cannot reference any attributes.")
   assert(serializer.flatMap { ser =>
     val boundRefs = ser.collect { case b: BoundReference => b }
-    assert(boundRefs.nonEmpty,
-      "each serializer expression should contain at least one `BoundReference`")
+    assert(boundRefs.nonEmpty || isEmptyStruct(ser),
```
Comment (milastdbx) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:328]: what if you have empty boundref in nonStruct and in struct as well ?

**Comment by cloud-fan** [comment]
Does it mean we will hit this bug if an empty struct is returned? `EXCEPT` may not be the only way to create empty struct.

**Comment by stefankandic** [comment]
> Does it mean we will hit this bug if an empty struct is returned? `EXCEPT` may not be the only way to create empty struct.

yes, i think we will @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 2 above)
Comment (stefankandic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:328]: can you maybe elaborate a bit? I'm not 100% sure what you mean

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 3]
```diff
@@ -325,11 +325,17 @@ case class ExpressionEncoder[T](
...
+      "each serializer expression should contain at least one `BoundReference` " +
+        "or be an empty struct")
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:330]: The error msg is not clear.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:328]: The change looks correct to me. If a column is StructType, the only reason that its serialize has no `BoundReference` is it's an empty struct.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 3 above)
Comment (stefankandic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: updated! let me know if you think it's clearer now

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 2 above)
Comment (milastdbx) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:328]: I'm just saying that `isEmptyyStruct` condition can mask any cases where boundRefs is empty and it shouldn't be if empty struct is present.

I'll let someone with more context decide whether this is ok

**Review by milastdbx — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 2 above)
Comment (stefankandic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:328]: i could be wrong but it seems to me that the empty struct will always have empty boundref

**Review by beliefer — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
@stefankandic Just to double check, up to which version should the changes be ported? The ticket SPARK-46539  points out 3.0.0, is it correct?

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @stefankandic and @cloud-fan @beliefer @milastdbx for review.

**Comment by MaxGekk** [comment]
@stefankandic The PR conflicts with `branch-3.5`. Please, open a separate PR for the branch and `branch-3.4` (if it is needed).

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @stefankandic and @MaxGekk and all. Unfortunately, this seems to break `master` branch. Could you make a follow-up first before backporting?

- https://github.com/apache/spark/actions/runs/7401236452/job/20136573641
- https://github.com/apache/spark/actions/runs/7401841987/job/20138486695
- https://github.com/apache/spark/actions/runs/7403100751/job/20142273016

```
[info] - selectExcept.sql_analyzer_test *** FAILED *** (136 milliseconds)
[info]   selectExcept.sql_analyzer_test
[info]   Expected "... (`tbl_view`, [id#x,[name#x,]data#x])
[info]         +- Pr...", but got "... (`tbl_view`, [id#x,[ name#x, ]data#x])
[info]         +- Pr..." Result did not match for query #9
[info]   SELECT * EXCEPT (data.f1, data.s2) FROM tbl_view (SQLQueryTestSuite.scala:902)
```

**Comment by dongjoon-hyun** [comment]
As of now, this failure blocks all PRs like the following.
- https://github.com/dongjoon-hyun/spark/actions/runs/7403216690/job/20142603391

**Comment by cloud-fan** [comment]
I'm fixing it at https://github.com/apache/spark/pull/44585

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan !

**Comment by stefankandic** [comment]
> @stefankandic The PR conflicts with `branch-3.5`. Please, open a separate PR for the branch and `branch-3.4` (if it is needed).

since this PR depends on a new feature introduced in [43843](https://github.com/apache/spark/pull/43843) there is no need to backport this

**Pull request closed without merging**